### PR TITLE
Improve Docker

### LINKED ISSUE #1717: Gogs docker don't support ED25519 keys
If you try to add an ed25519 user ssh key the web responds:
> Sorry, we're not able to verify your SSH key: ssh-keygen -l -f: line 1 too long: ssh-ed25519 AAAAC3NsdaP5lJHI1NTA5BBBBIHSi...

Could be cause a very old SSH server?

### PR DESCRIPTION
**Pull request by ogarcia**
- Updated to use Debian Jessie base image.
- Less intermediate containers.
- Uses supervisor to start process.
- Fix #1717, #1708 and probably #1313.

### CODE REVIEW

**Comment by Unknwon** [comment]
Thanks!

Comments are made inline.

File: Dockerfile [hunk 1]
```diff
@@ -1,54 +1,48 @@
...
-    apt-get autoremove && \
-    rm -rf /var/lib/apt/lists/*
-
-ENV GOROOT /goroot
-ENV GOPATH /gopath
-ENV PATH $PATH:$GOROOT/bin:$GOPATH/bin
-
-COPY . /gopath/src/github.com/gogits/gogs/
-WORKDIR /gopath/src/github.com/gogits/gogs/
+FROM debian:jessie
```
Comment (Unknwon) [inline Dockerfile]: Any reason changed to this?

File: docker/start.sh [hunk 2]
```diff
@@ -40,4 +37,4 @@ then
 fi
 
 chown -R git:git /data .
-exec su git -c "./gogs web"
+exec supervisord -c /app/gogs/docker/supervisord.conf
```
Comment (Unknwon) [inline docker/start.sh]: Why use supervisor simply ./gogs web would work?

File: Dockerfile (hunk 1 above)
Comment (ogarcia) [inline Dockerfile]: Yes. The ENV variables still remain in docker although are unneeded cause GOROOT and GOPATH will be deleted.

I think that could keep COPY line (`COPY . /gopath/src/github.com/gogits/gogs/`), cause the Docker image are build with DockerHub, in this case we can remove the git clone lines (16 and 17):
> git clone https://github.com/gogits/gogs.git \
>    /gopath/src/github.com/gogits/gogs && \

And can remove the `mkdir /gopath/src/github.com/gogits` in 14 cause the copy do the magic.

Comment (ogarcia) [inline Dockerfile]: Mmmm. I will do this changes and update te PR. ;)

**Comment by ogarcia** [comment]
Is better with the second change c92515ae5d221e9a4ceff1663233bb630701a88f, cause the git clone not is neccesary inside docker if you have a local copy of repo.

File: docker/start.sh (hunk 2 above)
Comment (ogarcia) [inline docker/start.sh]: Cause supervisor runs sshd too.

In docker you must run only one master process. In this case the master process is `supervisord` and it have `sshd` and `gogs` as child processes.

Ah, and this must fix #1499. :wink:

File: Dockerfile (hunk 1 above)
Comment (ogarcia) [inline Dockerfile]: Ah! If you say the FROM line:
`FROM debian:jessie`

This base distro is officially supported by docker team, and is most recent than `google/debian:wheezy`.

It fixes the issues reated with SSH and ED25519 keys. :smile:

**Comment by tboerger** [comment]
I would suggest to use alpine as a base image, that way the container gets really tiny. And instead of supervisor I would use s6 as this works really correct as PID1 within docker, and it doesn't rely on python which increases the image size as well. I have already done something like that for [gitea](https://github.com/dockhippie/gitea), my image is still pretty large as I have installed logstash-forwarder and other tools in my base image.

**Comment by tboerger** [comment]
Beside that you are not rotating the logs and you are not printing the logs to stdout which sucks.

**Comment by ogarcia** [comment]
@tboerger we could use Alpine, but the main target is that the docker can be *easy to use and modify*, on the other hand, supervisor is only one package cause python is installed in base system.

I don't understand the problem with log rotation? Supervisor rotates the log automagically and in stdout you see the main proccess log (supervisord).

**Comment by tboerger** [comment]
It is easy to use and modify with alpine. Just python bloats the image already massively... Just as an example, a full alpine image takes ~5mb. And where rotates supervisor all the logs? Beside that got supervisor issues as pid1 in docker.

**Comment by Unknwon** [comment]
Thanks your work again!

And close by https://github.com/gogits/gogs/pull/1730 

You can continue improving based on https://github.com/gogits/gogs/pull/1730 .

**Pull request closed without merging**